lib/livebook/notebook_manager.ex

```diff
@@ -0,0 +1,287 @@
+
+    state = %{state | recent_notebooks: recent_notebooks, starred_notebooks: starred_notebooks}
+    broadcast_changes(state, prev_state)
+    {:noreply, state, {:continue, :dump_state}}
+  end
+
+  def handle_cast({:add_starred_notebook, file, name}, state = prev_state) do
+    if Enum.any?(state.starred_notebooks, &(&1.file == file)) do
+      {:noreply, state}
+    else
+      starred_notebooks = [
+        %{file: file, name: name, added_at: DateTime.utc_now()} | state.starred_notebooks
+      ]
+
+      state = %{state | starred_notebooks: starred_notebooks}
+      broadcast_changes(state, prev_state)
+      {:noreply, state, {:continue, :dump_state}}
+    end
+  end
+
+  def handle_cast({:remove_starred_notebook, file}, state = prev_state) do
+    starred_notebooks = Enum.reject(state.starred_notebooks, &(&1.file == file))
+    state = %{state | starred_notebooks: starred_notebooks}
+    broadcast_changes(state, prev_state)
+    {:noreply, state, {:continue, :dump_state}}
+  end
+
+  def handle_cast({:remove_file_system, file_system_id}, state = prev_state) do
+    recent_notebooks = remove_notebooks_on_file_system(state.recent_notebooks, file_system_id)
+    starred_notebooks = remove_notebooks_on_file_system(state.starred_notebooks, file_system_id)
+    state = %{state | recent_notebooks: recent_notebooks, starred_notebooks: starred_notebooks}
+    broadcast_changes(state, prev_state)
+    {:noreply, state, {:continue, :dump_state}}
+  end
+
+  def handle_cast({:update_notebook_name, file, name}, state = prev_state) do
+    recent_notebooks = update_notebook_names(state.recent_notebooks, file, name)
+    starred_notebooks = update_notebook_names(state.starred_notebooks, file, name)
+    state = %{state | recent_notebooks: recent_notebooks, starred_notebooks: starred_notebooks}
+    broadcast_changes(state, prev_state)
+    {:noreply, state, {:continue, :dump_state}}
+  end
+
+  defp remove_notebooks_on_file_system(notebook_infos, file_system_id) do
+    Enum.reject(notebook_infos, &(&1.file.file_system.id == file_system_id))
+  end
+
+  defp update_notebook_names(notebook_infos, file, name) do
+    Enum.map(notebook_infos, fn
+      %{file: ^file} = info -> %{info | name: name}
+      info -> info
+    end)
+  end
+
+  defp broadcast_changes(state, prev_state) do
+    if state.recent_notebooks != prev_state.recent_notebooks do
+      Phoenix.PubSub.broadcast(
+        Livebook.PubSub,
+        "notebook_manager:recent_notebooks",
+        {:recent_notebooks_updated, state.recent_notebooks}
+      )
+    end
+
+    if state.starred_notebooks != prev_state.starred_notebooks do
+      Phoenix.PubSub.broadcast(
+        Livebook.PubSub,
+        "notebook_manager:starred_notebooks",
+        {:starred_notebooks_updated, state.starred_notebooks}
+      )
+    end
+  end
+
+  defp load_state() do
+    attrs =
+      case Storage.fetch(@namespace, @entity_id) do
```

Comment:
Maybe one namespace for starred and another for recent? Anyway is fine.

Reply:
Attributes can be written/read separately anyway, and this way we can fetch the whole state at once :)